### PR TITLE
added checkpoint meta object and saved meta information in checkpoint file names

### LINKED ISSUE #4: Checkpoint names and retrieval
We need to retrieve checkpoints by their metadata (e.g. some name, age, generation, tag).
If we store the metadata in the filename, we do not need to open the remote `{{CHECKPOINT_NAME}}.json` to retrieve the checkpoint. 

Questions: 
1. Which metadata should be shown to the user about his checkpoints?
2. Which metadata should identify a checkpoint and allow the user to retrieve it?
3. Should the metadata be stored inside the filename?

### PR DESCRIPTION
**Pull request by eichenroth**
fixes #4

### CODE REVIEW

File: pyback/checkpoint.py [hunk 1]
```diff
@@ -96,12 +97,19 @@ def from_dict(d):
 
 
 class Checkpoint:
-    def __init__(self, root, time=None):
+    def __init__(self, root, time=None, name=None):
         self.root = root
         self.time = datetime.now() if time is None else time
+        
+        assert name is None or re.match('^[a-zA-Z0-9_\-:.]+$', name)
+        self.name = self.time.isoformat() if name is None else name
```
Comment (klaudworks) [inline pyback/checkpoint.py]: Why do you set the name to the time?

File: pyback/checkpoint.py [hunk 2]
```diff
@@ -114,13 +122,30 @@ def iter(self):
...
+    def __init__(self, checksum, time, name):
+        self.checksum = checksum
+        self.time = time
+        self.name = name
+
+    def to_string(self):
+        timestring = self.time.isoformat()
+        timestring += '.000000' if len(timestring) == 19 else ''
+
+        return self.checksum + '_' + timestring + '_' + self.name
```
Comment (klaudworks) [inline pyback/checkpoint.py]: i.e. HASH_time_time if no name is set?

File: pyback/checkpoint.py [hunk 3]
```diff
@@ -114,13 +122,30 @@ def iter(self):
...
+
+    @staticmethod
+    def from_string(string):
+        return CheckpointMeta(string[:16], datetime_from_iso_format(string[17:43]), string[44:])
```
Comment (klaudworks) [inline pyback/checkpoint.py]: I would prefer a split on the underscores. You can specify that it should split 3 times so you do not interfere with the underscores in the name.

File: pyback/storage.py [hunk 4]
```diff
@@ -85,26 +92,44 @@ def retrieve_file(self, checksum, dst_file_path, file_permissions):
...
         if not os.path.exists(os.path.dirname(tree_file_path)):
             os.makedirs(os.path.dirname(tree_file_path))
         with open(tree_file_path, 'w') as f:
             f.write(checkpoint.to_json())
 
-    # TODO We should probably also retrieve some metadata about the checkpoint
-    #  except from the name to allow the user to chose between checkpoints depending on a timestamp, name etc.
-    def retrieve_checkpoint_ids(self):
-        return [checkpoint[:-len(self.TREE_FILE_EXT)] for checkpoint in os.listdir(self.tree_path) if checkpoint[-len(self.TREE_FILE_EXT):] == self.TREE_FILE_EXT]
+    def retrieve_checkpoint_metas(self):
```
Comment (klaudworks) [inline pyback/storage.py:111]: this is/was really hard to read for me

File: pyback/storage.py [hunk 5]
```diff
@@ -85,26 +92,44 @@ def retrieve_file(self, checksum, dst_file_path, file_permissions):
...
+            if checkpoint_file in checkpoint_files:
+                return checkpoint_from_json_file(checkpoint_file)
+        elif isinstance(checkpoint_id, str):
+            # checksum
+            for checkpoint_file in checkpoint_files:
+                if checkpoint_file[:len(checkpoint_id)] == checkpoint_id:
+                    return checkpoint_from_json_file(checkpoint_file)
+            # name
+            for checkpoint_file in checkpoint_files:
+                if checkpoint_file[44:-len(self.TREE_FILE_EXT)] == checkpoint_id:
```
Comment (klaudworks) [inline pyback/storage.py]: see above, I think a split on the underscores would be easier to read

File: pyback/storage.py [hunk 6]
```diff
@@ -85,26 +92,44 @@ def retrieve_file(self, checksum, dst_file_path, file_permissions):
...
+                            if c[-len(self.TREE_FILE_EXT):] == self.TREE_FILE_EXT]
+
+        # checkpoint meta
+        if isinstance(checkpoint_id, CheckpointMeta):
+            checkpoint_file = checkpoint_id.to_string() + self.TREE_FILE_EXT
+            if checkpoint_file in checkpoint_files:
+                return checkpoint_from_json_file(checkpoint_file)
+        elif isinstance(checkpoint_id, str):
+            # checksum
+            for checkpoint_file in checkpoint_files:
```
Comment (klaudworks) [inline pyback/storage.py]: If we match the first checkpoint file with a matching hash, we should ensure that the checkpoints are read in reversed time order.

File: pyback/checkpoint.py (hunk 1 above)
Comment (klaudworks) [reply pyback/checkpoint.py]: We could just leave the name out in this case.

**Review by klaudworks — CHANGES_REQUESTED** [review]

File: pyback/storage.py (hunk 5 above)
Comment (eichenroth) [reply pyback/storage.py]: Shouldn't we throw an error if the hash prefix is ambiguous?

**Pull request merged**